go/database/mpt/verification.go

```diff
@@ -570,6 +586,73 @@ func verifyHashesStoredWithParents[N any](
 	return nil
 }
 
+// verifyContractCodes runs list of validation checks on Mpt and on the
+// code file stored in the given directory. These checks include:
+// 1) Fatal checks
+// - All CodeHashes within the Mpt are present in the code file
+// - All CodeHashes within the code file are correct
+// 2) Non-fatal checks
+// - There are no extra Code Hashes not referenced by any account
+func verifyContractCodes(directory string, source *verificationNodeSource, observer VerificationObserver) error {
+	observer.Progress(fmt.Sprintf("Checking contract codes ..."))
+
+	codeFile := directory + "/codes.dat"
+	// make sure the file exists
+	if _, err := os.Stat(codeFile); err != nil {
+		return fmt.Errorf("code file %v does not exist", codeFile)
+	}
+	codes, err := readCodes(codeFile)
+	if err != nil {
+		return err
+	}
+
+	found := make(map[common.Hash]bool)
+	check := func(acc *AccountNode) error {
+		accountHash := acc.info.CodeHash
+		// skip accounts that are not contracts
+		if accountHash == emptyCodeHash {
+			return nil
+		}
+		// check that the code hash is present in the code file
+		byteCode, exists := codes[accountHash]
+		if !exists {
+			return fmt.Errorf("the hash %x is present in the mpt but missing from the code file", accountHash)
+		}
+		// check correctness of the code hash
+		if got, want := common.Keccak256(byteCode), &accountHash; got.Compare(want) != 0 {
+			return fmt.Errorf("unexpected code hash for address, got: %x want: %x", got, want)
+		}
```

Comment:
This should be done outside the Account Node iteration. If a code is used 100x, the same code is checked 100 times.

Reply:
I implemented an early exit. So we dont need to create a second loop.